contrib/TestHarness2/test_harness/summarize.py

```diff
@@ -0,0 +1,653 @@
+        else:
+            return False
+
+    def __ge__(self, other):
+        if isinstance(other, tuple) and len(other) == 3:
+            return self.to_tuple() >= other
+        elif isinstance(other, Coverage):
+            return self.to_tuple() >= other.to_tuple()
+        else:
+            return False
+
+    def __hash__(self):
+        return hash((self.file, self.line, self.comment))
+
+
+class TraceFiles:
+    def __init__(self, path: Path):
+        self.path: Path = path
+        self.timestamps: List[int] = []
+        self.runs: OrderedDict[int, List[Path]] = collections.OrderedDict()
+        trace_expr = re.compile(r'trace.*\.(json|xml)')
+        for file in self.path.iterdir():
+            if file.is_file() and trace_expr.match(file.name) is not None:
+                ts = int(file.name.split('.')[6])
+                if ts in self.runs:
+                    self.runs[ts].append(file)
+                else:
+                    self.timestamps.append(ts)
+                    self.runs[ts] = [file]
+        self.timestamps.sort(reverse=True)
+
+    def __getitem__(self, idx: int) -> List[Path]:
+        res = self.runs[self.timestamps[idx]]
+        res.sort()
+        return res
+
+    def __len__(self) -> int:
+        return len(self.runs)
+
+    def items(self) -> Iterator[List[Path]]:
+        class TraceFilesIterator(Iterable[List[Path]]):
+            def __init__(self, trace_files: TraceFiles):
+                self.current = 0
+                self.trace_files: TraceFiles = trace_files
+
+            def __iter__(self):
+                return self
+
+            def __next__(self) -> List[Path]:
+                if len(self.trace_files) <= self.current:
+                    raise StopIteration
+                self.current += 1
+                return self.trace_files[self.current - 1]
+        return TraceFilesIterator(self)
+
+
+class Summary:
+    def __init__(self, binary: Path, runtime: float = 0, max_rss: int | None = None,
+                 was_killed: bool = False, uid: uuid.UUID | None = None, expected_unseed: int | None = None,
+                 exit_code: int = 0, valgrind_out_file: Path | None = None, stats: str | None = None,
+                 error_out: str = None, will_restart: bool = False):
+        self.binary = binary
+        self.runtime: float = runtime
+        self.max_rss: int | None = max_rss
+        self.was_killed: bool = was_killed
+        self.expected_unseed: int | None = expected_unseed
+        self.exit_code: int = exit_code
+        self.out: SummaryTree = SummaryTree('Test')
+        self.test_begin_found: bool = False
+        self.test_end_found: bool = False
+        self.unseed: int | None = None
+        self.valgrind_out_file: Path | None = valgrind_out_file
+        self.severity_map: OrderedDict[tuple[str, int], int] = collections.OrderedDict()
+        self.error: bool = False
+        self.errors: int = 0
+        self.error_list: List[str] = []
```

Comment:
What is `error_list` used for? I see some items being added to it, but I didn't spot where the items were being read.

Reply:
This was me rewriting the old test harness and not knowing what this is used for (so I just started building the list). I turns out before we were writing this to a separate xml file but I can't find a use for this. I am going to remove this field.